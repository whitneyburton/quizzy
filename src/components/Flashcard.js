import React, { Component } from 'react';
import '../styles/main.scss';

export default class Flashcard extends Component {
  constructor() {
    super();
    this.state = {
      correct: null
    }
  }

  validateAnswer = (e) => {
    let { flashcard, saveToStorage, removeCorrectFromStorage } = this.props;
    let answerClicked = e.target.innerText;
    if (flashcard.answer === answerClicked) {
      this.setState({ correct: true });
      removeCorrectFromStorage(flashcard);
    } else {
      this.setState({ correct: false });
      saveToStorage(flashcard);
    }
  }

  randomizeAnswers() {
    let { flashcard, flashcards } = this.props;
    let allAnswers = flashcards
      .map(flashcard => flashcard.answer)
      .sort(() => 0.5 - Math.random())
    let correctAnswerIndex = allAnswers.indexOf(flashcard.answer)
    allAnswers.splice(correctAnswerIndex, 1)
    let finalArray = [...allAnswers
      .splice(0, 2), flashcard.answer]
      .sort(() => 0.5 - Math.random())
    return finalArray;
  }

  render() {
    const { flashcard, filteredCards } = this.props;
    const { correct } = this.state;
    let currentAnswersArray = this.randomizeAnswers();
    let flashcardClass = 'Flashcard';
    let correctFeedback;
    let incorrectFeedback;
    if (correct === true) {
      flashcardClass = 'Flashcard correct-answer'
      correctFeedback = 'show'
      incorrectFeedback = 'hide'
    } else if (correct === false) {
      flashcardClass = 'Flashcard incorrect-answer'
      incorrectFeedback = 'show'
      correctFeedback = 'hide'
    } else {
      incorrectFeedback = 'hide'
      correctFeedback = 'hide'
    }

    return (
      <div className={flashcardClass}>
        <p className="question-counter">
          Question {filteredCards.indexOf(flashcard) + 1}/{filteredCards.length}
        </p>
        <p className="flashcard-question">{flashcard.question}</p>
        <div className="buttons-container">
          {currentAnswersArray.map((answer, index) => {
            return <button
              key={index}
              onClick={this.validateAnswer}
              className="buttons flashcard-buttons"
              type="button">{answer}</button>
          })}
        </div>
        <p className={correctFeedback}>That's right! Nice work, Quizzy pro.</p>
        <p className={incorrectFeedback}>Not quite. Keep guessing!</p>
        <div className="learn-more">
          <a
            href={flashcard.mdn_link}
            className="mdn-link"
            target="_blank"
            rel="noopener noreferrer">Learn More</a>
        </div>
      </div>
    )
  }
}