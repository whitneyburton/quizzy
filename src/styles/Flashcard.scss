.Flashcard {
  border: 1px solid black;
  border-radius: 5px;
  background-color: #ccdedc;
  width: 75%;
  // height: 5000px;
  margin: 10px 0px;
  padding: 28px;
}

.question-counter {
  text-align: right;
  font-size: 12px;
}

.flashcard-question {
  margin: 20px 0px;
}

.buttons-container {
  display: flex;
  justify-content: space-evenly;
}

.flashcard-buttons {
  padding: 10px;
}

.hide {
  display: none;
}