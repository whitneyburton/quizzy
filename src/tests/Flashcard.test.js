import React from 'react';
import Flashcard from '../Components/Flashcard';
import { shallow } from 'enzyme';

describe('Flashcard', () => {
  let wrapper;
  const id = 1;
  const flashcard = {};
  const filteredCards = [{}, {}, {}]
  const saveToStorageMock = jest.fn();

  beforeEach(() => {
    wrapper = shallow(
      <Flashcard
        key={id}
        flashcard={flashcard}
        filteredCards={filteredCards}
        saveToStorage={saveToStorageMock}
      />);
  });

  it('should match the snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  })
})