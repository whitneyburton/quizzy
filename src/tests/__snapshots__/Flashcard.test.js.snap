// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Flashcard should match the snapshot 1`] = `
<div
  className="Flashcard"
>
  <p
    className="question-counter"
  >
    Question 
    0
    /
    3
  </p>
  <p
    className="flashcard-question"
  />
  <div
    className="buttons-container"
  >
    <button
      className="buttons flashcard-buttons"
      key="0"
      onClick={[Function]}
      type="button"
    >
      .map()
    </button>
    <button
      className="buttons flashcard-buttons"
      key="1"
      onClick={[Function]}
      type="button"
    >
      .reduce()
    </button>
    <button
      className="buttons flashcard-buttons"
      key="2"
      onClick={[Function]}
      type="button"
    >
      .copyWithin()
    </button>
  </div>
  <p
    className="hide"
  >
    That's right! Nice work, Quizzy pro.
  </p>
  <p
    className="hide"
  >
    Not quite. Keep guessing!
  </p>
  <div
    className="learn-more"
  >
    <a
      className="mdn-link"
      rel="noopener noreferrer"
      target="_blank"
    >
      Learn More
    </a>
  </div>
</div>
`;
